Jenkinsfile

```diff
@@ -0,0 +1,139 @@
+            environment {
+                ENVIRONMENT = defineEnvironmentByBranchName()
+            }
+            steps {
+                sh '''docker build \
+                    -t "relay-webapp::${BRANCH_NAME}-commit-${GIT_COMMIT}" \\
+                    --build-arg environment=${ENVIRONMENT} \\
+                    --build-arg version=${ENVIRONMENT}-${BRANCH_NAME}+${GIT_COMMIT} \\
+                    -f Dockerfile \\
+                    .'''
+            }
+        }
+        stage('Test') {
+            steps {
+                sh 'docker build --target test .'
+            }
+        }
+
+        stage('Delivery docker images') {
+            when { 
+                anyOf { 
+                    branch 'master'
+                    branch 'main'
+                    branch 'develop'
+                    branch 'qa'
+                    buildingTag()
+                }
+            }            
+            environment {
+                DOCKER_NAMESPACE = 'dopplerdock'
+            }
+            steps {
+                sh 'docker tag relay-webapp:${BRANCH_NAME}-commit-${GIT_COMMIT} relay-webapp:${BRANCH_NAME}'
+                withDockerRegistry([ credentialsId: "dockerhub_${DOCKER_NAMESPACE}", url: ""]) {
+                    sh 'docker push relay-webapp:${BRANCH_NAME}'
+                }
+            }
+        }
+
+        stage('Delivery semver docker images') {
+            when { 
+                expression {
+                    return isVersionTag(readCurrentTag())
+                }
+            }
+            environment {
+                DOCKER_NAMESPACE = 'dopplerdock'
+                SEMVER_MAYOR = "${env.BRANCH_NAME.tokenize('.')[0]}"
+                SEMVER_MAYOR_MINOR = "${SEMVER_MAYOR}.${env.BRANCH_NAME.tokenize('.')[1]}"
+                SEMVER_MAYOR_MINOR_PATCH = "${SEMVER_MAYOR_MINOR}.${env.BRANCH_NAME.tokenize('.')[2]}"
+            }
+            steps {
+                sh 'docker tag relay-webapp:${BRANCH_NAME}-commit-${GIT_COMMIT} relay-webapp:${SEMVER_MAYOR}'
+                sh 'docker tag relay-webapp:${BRANCH_NAME}-commit-${GIT_COMMIT} relay-webapp:${SEMVER_MAYOR_MINOR}'
+                sh 'docker tag relay-webapp:${BRANCH_NAME}-commit-${GIT_COMMIT} relay-webapp:${SEMVER_MAYOR_MINOR_PATCH}'
+                sh 'docker tag relay-webapp:${BRANCH_NAME}-commit-${GIT_COMMIT} relay-webapp:latest'
+                
+                withDockerRegistry([ credentialsId: "dockerhub_${DOCKER_NAMESPACE}", url: ""]) {
+                    sh 'docker push relay-webapp:${SEMVER_MAYOR}'
+                    sh 'docker push relay-webapp:${SEMVER_MAYOR_MINOR}'
+                    sh 'docker push relay-webapp:${SEMVER_MAYOR_MINOR_PATCH}'
+                    sh 'docker push relay-webapp:latest'
+                }
+            }
+        }
+    }
+    post { 
+        cleanup { 
+            cleanWs()
+            dir("${env.WORKSPACE}@tmp") {
+                deleteDir()
+            }
+        }
+    }
```

Comment:
Is it necessary?

Reply:
In the web app, it is necessary because we are building the files out of the docker image.

In general, we are trying to avoid doing that.